docs/doxygen/man/man3/GetMessage.3
.TH "GetMessage" 3 "Thu Aug 30 2018" "esc" \" -*- nroff -*-
.ad l
.nh
.SH NAME
GetMessage \- Get message command class\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <getmessage\&.h>\fP
.PP
Inherits \fBBlockCommand\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBGetMessage\fP (uint16_t abank, uint32_t auser, uint32_t block, uint16_t dstNode, uint32_t msgId, uint32_t time)"
.br
.ti -1c
.RI "virtual int \fBgetType\fP () override"
.br
.RI "\fIReturn TXSTYPE_MSG as type \&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetData\fP () override"
.br
.RI "\fIGet pointer to command data structure\&. \fP"
.ti -1c
.RI "virtual CommandType \fBgetCommandType\fP () override"
.br
.RI "\fIReturn eReadingOnly as command type \&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetResponse\fP () override"
.br
.RI "\fIGet pointer to response data\&. \fP"
.ti -1c
.RI "virtual void \fBsetData\fP (char *data) override"
.br
.RI "\fIPut data as a char list and put convert it to data structure\&. \fP"
.ti -1c
.RI "virtual void \fBsetResponse\fP (char *response) override"
.br
.RI "\fIApply data as a response struct\&. \fP"
.ti -1c
.RI "virtual int \fBgetDataSize\fP () override"
.br
.RI "\fIGet data struct size\&. Without signature\&. \fP"
.ti -1c
.RI "virtual int \fBgetResponseSize\fP () override"
.br
.RI "\fIGet response data struct size\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetSignature\fP () override"
.br
.RI "\fIGet pointer to signature data\&. \fP"
.ti -1c
.RI "virtual int \fBgetSignatureSize\fP () override"
.br
.RI "\fIGet signature size\&. \fP"
.ti -1c
.RI "virtual void \fBsign\fP (const uint8_t *hash, const uint8_t *sk, const uint8_t *pk) override"
.br
.RI "\fISign actual data plus hash using user private and public keys\&. \fP"
.ti -1c
.RI "virtual bool \fBcheckSignature\fP (const uint8_t *hash, const uint8_t *pk) override"
.br
.RI "\fICheck actual signature\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetTime\fP () override"
.br
.RI "\fIGet time of command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserId\fP () override"
.br
.RI "\fIGet User ID\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetBankId\fP () override"
.br
.RI "\fIGet Node ID\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetFee\fP () override"
.br
.RI "\fIGet command fee\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetDeduct\fP () override"
.br
.RI "\fIGet change in cash balance after command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserMessageId\fP () override"
.br
.RI "\fIGet message id\&. \fP"
.ti -1c
.RI "virtual bool \fBsend\fP (\fBINetworkClient\fP &netClient) override"
.br
.RI "\fISend data to the server\&. \fP"
.ti -1c
.RI "virtual void \fBsaveResponse\fP (\fBsettings\fP &sts) override"
.br
.RI "\fISave command response to settings object\&. \fP"
.ti -1c
.RI "virtual std::string \fBtoString\fP (bool pretty) override"
.br
.ti -1c
.RI "virtual void \fBtoJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual void \fBtxnToJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual std::string \fBusageHelperToString\fP () override"
.br
.ti -1c
.RI "uint32_t \fBgetBlockTime\fP ()"
.br
.ti -1c
.RI "uint16_t \fBgetDestNode\fP ()"
.br
.ti -1c
.RI "uint32_t \fBgetMsgId\fP ()"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "\fBGetMessageData\fP \fBm_data\fP"
.br
.ti -1c
.RI "\fBcommandresponse\fP \fBm_response\fP"
.br
.ti -1c
.RI "message_ptr \fBm_responseMsg\fP"
.br
.ti -1c
.RI "uint32_t \fBm_responseBlock\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Get message command class\&. 

TODO: not finished class 
.PP
Definition at line 11 of file getmessage\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "bool GetMessage::checkSignature (const uint8_t * hash, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Check actual signature\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 64 of file getmessage\&.cpp\&.
.SS "bool GetMessage::send (\fBINetworkClient\fP & netClient)\fC [override]\fP, \fC [virtual]\fP"

.PP
Send data to the server\&. 
.PP
\fBParameters:\fP
.RS 4
\fInetClient\fP Netwrok client implementation of \fBINetworkClient\fP interface\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 91 of file getmessage\&.cpp\&.
.SS "void GetMessage::sign (const uint8_t * hash, const uint8_t * sk, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Sign actual data plus hash using user private and public keys\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIsk\fP Pointer to provate key\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 60 of file getmessage\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for esc from the source code\&.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2018-08-24
### Changed
- Genesis file for the launch
- Update README

## [0.0.6] - 2018-08-10
### Added
- Profit sharing among TOP and VIP nodes

### Changed
- Genesis file for test launch

### Fixed
- Vote counting in get_block command
- Ability to send externally signed transactions

## [0.0.5] - 2018-08-01
### Added
- Support passing version by cmake command.
- Contributing guide
- "Dry run" test case

### Fixed
- Node stops after incorrect `send_many` request  - #134
- Fee checking
- Internal transactions numbers

## [0.0.4] - 2018-07-27
### Changed
- Update default parameters
- Creating a dev version with reference to the last tag

[Unreleased]: https://github.com/adshares/ads/compare/v1.0.0...HEAD

[1.0.0]: https://github.com/adshares/ads/compare/v0.0.6...v1.0.0
[0.0.6]: https://github.com/adshares/ads/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/adshares/ads/compare/v0.0.3...v0.0.4


src/common/command/getmessage.cpp
#include "getmessage.h"

#include "ed25519/ed25519.h"
#include "abstraction/interfaces.h"
#include "helper/json.h"
#include "helper/hash.h"
#include "message.hpp"
#include "command/factory.h"

GetMessage::GetMessage()
    : m_data{} {
    m_responseError = ErrorCodes::Code::eNone;
}

GetMessage::GetMessage(uint16_t abank, uint32_t auser, uint32_t block, uint16_t dstNode, uint32_t msgId, uint32_t time)
    : m_data(abank, auser, block, dstNode, msgId, time) {
    m_responseError = ErrorCodes::Code::eNone;
}

unsigned char* GetMessage::getData() {
    return reinterpret_cast<unsigned char*>(&m_data.info);
}

unsigned char* GetMessage::getResponse() {
    return reinterpret_cast<unsigned char*>(&m_response);
}

void GetMessage::setData(char* data) {
    m_data = *reinterpret_cast<decltype(m_data)*>(data);
}

void GetMessage::setResponse(char* response) {
    m_response = *reinterpret_cast<decltype(m_response)*>(response);
}

int GetMessage::getDataSize() {
    return sizeof(m_data.info);
}

int GetMessage::getResponseSize() {
    return sizeof(m_response);
}

unsigned char* GetMessage::getSignature() {
    return m_data.sign;
}

int GetMessage::getSignatureSize() {
    return sizeof(m_data.sign);
}

int GetMessage::getType() {
    return TXSTYPE_MSG;
}

CommandType GetMessage::getCommandType() {
    return CommandType::eReadingOnly;
}

void GetMessage::sign(const uint8_t* /*hash*/, const uint8_t* sk, const uint8_t* pk) {
    ed25519_sign(getData(), getDataSize(), sk, pk, getSignature());
}

bool GetMessage::checkSignature(const uint8_t* /*hash*/, const uint8_t* pk) {
    return (ed25519_sign_open(getData(), getDataSize(), pk, getSignature()) == 0);
}

void GetMessage::saveResponse(settings& /*sts*/) {
}

uint32_t GetMessage::getUserId() {
    return m_data.info.src_user;
}

uint32_t GetMessage::getBankId() {
    return m_data.info.src_node;
}

uint32_t GetMessage::getTime() {
    return m_data.info.ttime;
}

int64_t GetMessage::getFee() {
    return 0;
}

int64_t GetMessage::getDeduct() {
    return 0;
}

bool GetMessage::send(INetworkClient& netClient) {
    if(!netClient.sendData(getData(), sizeof(m_data))) {
        ELOG("GetMessage sending error\n");
        return false;
    }

    if (!netClient.readData((int32_t*)&m_responseError, ERROR_CODE_LENGTH)) {
        ELOG("GetMessage reading error\n");
        return false;
    }

    if (m_responseError) {
        return true;
    }

    if (!netClient.readData((int32_t*)&m_responseBlock, sizeof(uint32_t))) {
        ELOG("GetMessage reading message block\n");
        return false;
    }

    // msgTypeLength contains type (1B) and length (3B)
    uint32_t msgTypeLength = 0;
    if (!netClient.readData((int32_t*)&msgTypeLength, sizeof(msgTypeLength))) {
        ELOG("GetMessage reading length error\n");
        return false;
    }
    uint32_t msgLength = msgTypeLength >> 8;
    if (!msgTypeLength || msgLength < message::data_offset) {
        m_responseError = ErrorCodes::Code::eBadLength;
        return true;
    }
    uint8_t msgType = msgTypeLength & 0xFF;
    if (msgType != MSGTYPE_MSG) {
        m_responseError = ErrorCodes::Code::eIncorrectTransaction;
        return true;
    }

    char buffer[msgLength];
    memcpy(buffer, &msgTypeLength, 4);
    if (!netClient.readData(buffer + 4, msgLength - 4)) {
        ELOG("GetMessage reading error\n");
        return false;
    }

    m_responseMsg.reset(new message(msgLength, (uint8_t*)buffer));

    return true;
}

uint32_t GetMessage::getBlockTime() {
    return m_data.info.block;
}

uint16_t GetMessage::getDestNode() {
    return m_data.info.dst_node;
}

uint32_t GetMessage::getMsgId() {
    return m_data.info.node_msgid;
}

uint32_t GetMessage::getUserMessageId() {
    return 0;
}

std::string GetMessage::toString(bool /*pretty*/) {
    return "";
}

void GetMessage::toJson(boost::property_tree::ptree& ptree) {
    if (m_responseError) {
        ptree.put(ERROR_TAG, ErrorCodes().getErrorMsg(m_responseError));
    } else {
        m_responseMsg->read_head();

        char blockhex[9];
        blockhex[8]='\0';
        sprintf(blockhex,"%08X",m_responseBlock);

        ptree.put("block_id", blockhex);
        ptree.put("message_id",Helper::print_msg_pack_id(m_responseMsg->svid, m_responseMsg->msid));
        ptree.put("node", m_responseMsg->svid);
        ptree.put("node_msid", m_responseMsg->msid);
        ptree.put("time", m_responseMsg->now);
        ptree.put("length", m_responseMsg->len);
        if(!m_responseMsg->hash_tree_fast(m_responseMsg->sigh,m_responseMsg->data,m_responseMsg->len,m_responseMsg->svid,m_responseMsg->msid)) {
            return;
        }
        ptree.put("hash", Helper::ed25519_key2text(m_responseMsg->sigh, sizeof(m_responseMsg->sigh)));
        uint8_t* data_ptr = m_responseMsg->data+message::data_offset;
        uint8_t* data_end = m_responseMsg->data+m_responseMsg->len;
        uint32_t mpos = 1;
        boost::property_tree::ptree transactions;
        std::unique_ptr<IBlockCommand> command;
        while (data_ptr < data_end) {
            command = command::factory::makeCommand(*data_ptr);
            if (command) {
                boost::property_tree::ptree txn;
                command->setData((char*)data_ptr);
                int size = command->getBlockMessageSize();

                txn.put("id", Helper::print_msg_id(m_responseMsg->svid, m_responseMsg->msid, mpos++));
                command->txnToJson(txn);
                txn.put("size", size);

                data_ptr += size;
                transactions.push_back(std::make_pair("", txn));
            } else {
                m_responseError = ErrorCodes::Code::eIncorrectTransaction;
                ptree.put(ERROR_TAG, ErrorCodes().getErrorMsg(m_responseError));
                return;
            }
        }
        ptree.add_child("transactions", transactions);
    }
}

void GetMessage::txnToJson(boost::property_tree::ptree& ptree) {
    using namespace Helper;
    ptree.put(TAG::TYPE, getTxnName(m_data.info.ttype));
    ptree.put(TAG::SRC_NODE, m_data.info.src_node);
    ptree.put(TAG::SRC_USER, m_data.info.src_user);
    ptree.put(TAG::TIME, m_data.info.ttime);
    ptree.put(TAG::BLOCK, m_data.info.block);
    ptree.put(TAG::DST_NODE, m_data.info.dst_node);
    ptree.put(TAG::NODE_MSGID, m_data.info.node_msgid);
    ptree.put(TAG::SIGN, ed25519_key2text(getSignature(), getSignatureSize()));
}

std::string GetMessage::usageHelperToString() {
    std::stringstream ss{};
    ss << "Usage: " << "{\"run\":\"get_message\",\"message_id\":<message_id>,[\"block\":<block_id as hex>]}" << "\n";
    ss << "Example: " << "{\"run\":\"get_message\",\"message_id\":\"0001:0000000E\",\"block\":\"5B0573C0\"}" << "\n";
    return ss.str();
}
